probe_src/probe_frontend/src/ops.rs

```diff
@@ -0,0 +1,410 @@
+#[allow(unused_imports)]
+pub use crate::ffi::{
+    dev_t, gid_t, ino_t, mode_t, rusage, statx, statx_timestamp, timespec, timeval, uid_t, CloneOp,
+    CloseOp, ExitOp, GetRUsageOp, InitProcessOp, InitThreadOp, WaitOp,
+};
+use color_eyre::eyre::{eyre, Context};
+pub use std::ffi::{c_int, c_uint};
+
+use color_eyre::eyre::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    ffi::{OsStr, OsString},
+    os::unix::ffi::OsStrExt,
+    slice,
+};
+
+use crate::{arena::ArenaContext, ffi};
+
+/// Specialized version of [`std::convert::From`] for working with libprobe arena structs.
+///
+/// Since [`ffi`] structs from arena allocator files have intrinsically invalid pointers (because
+/// they came from a different virtual memory space). This trait and It's sibling [`FfiInto`]
+/// exist to act as [`From`] and [`Into`] with an added parameter of a [`ArenaContext`] that can be
+/// used to decode pointers.
+pub(crate) trait FfiFrom<T> {
+    fn ffi_from(value: &T, ctx: &ArenaContext) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+/// Specialized version of [`std::convert::Into`] for working with libprobe arena structs.
+///
+/// Much like [`std::convert::Into`] this trait is implemented automatically with a blanket
+/// implementation as the reciprocal of [`FfiFrom`].
+pub(crate) trait FfiInto<T> {
+    fn ffi_into(&self, ctx: &ArenaContext) -> Result<T>;
+}
+
+impl<T, U> FfiInto<U> for T
+where
+    U: FfiFrom<T>,
+{
+    #[inline]
+    fn ffi_into(&self, ctx: &ArenaContext) -> Result<U> {
+        U::ffi_from(self, ctx)
+    }
+}
+
+fn try_to_osstring(str: *const i8, ctx: &ArenaContext) -> Result<OsString> {
+    log::debug!("[unsafe] Parsing arena pointer: {:#x}", str as usize);
+    Ok(if str.is_null() {
+        OsString::new()
+    } else {
+        match ctx.try_deref(str as usize) {
+            Some(x) => {
+                OsStr::from_bytes(unsafe { slice::from_raw_parts(x, libc::strlen(x as *const i8)) })
+                    .to_os_string()
+            }
+            None => return Err(eyre!("Unable to lookup pointer {0:#x}", (str as usize))),
+        }
+    })
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Path {
```

Comment:
What's the diff between this file and `bindings.rs`? Can it be automatically generated?

Reply:
Except for possibly writing a proc-macro, there's no way I was able to figure out to automatically convert Ops containing pointers to versions with some rust string type. Since the pointers need special handling and the resultant type needs to be serializable, I can't just leave the pointers as-is.